lib/kubernetes-deploy/render_task.rb

```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+require 'tempfile'
+
+require 'kubernetes-deploy/renderer'
+require 'kubernetes-deploy/template_discovery'
+
+module KubernetesDeploy
+  class RenderTask
+    def initialize(logger:, current_sha:, template_dir:, bindings:)
+      @logger = logger
+      @template_dir = template_dir
+      @renderer = KubernetesDeploy::Renderer.new(
+        current_sha: current_sha,
+        bindings: bindings,
+        template_dir: @template_dir,
+        logger: @logger,
+      )
+    end
+
+    def run(*args)
+      run!(*args)
+      true
+    rescue KubernetesDeploy::FatalDeploymentError
+      false
+    end
+
+    def run!(stream, only_filenames = [])
+      @logger.reset
+      @logger.phase_heading("Initializing render task")
+
+      filenames = if only_filenames.empty?
+        TemplateDiscovery.new(@template_dir).templates
+      else
+        only_filenames
+      end
+      exceptions = []
+      validate_configuration(filenames)
+
+      @logger.phase_heading("Rendering template(s)")
+
+      filenames.each do |filename|
+        begin
+          render_filename(filename, stream)
+        rescue KubernetesDeploy::InvalidTemplateError => exception
+          exceptions << exception
+          log_invalid_template(filename, exception)
+        rescue KubernetesDeploy::FatalDeploymentError => exception
+          exceptions << exception
+          log_render_failure(exception)
```

Comment:
Why have a separate case for invalid template error above? It would be possible to take advantage of the augmented information on `InvalidTemplateError` to improve the message, `log_invalid_template` doesn't actually do that. More concerning, I don't think this case has test coverage (this line is missing the `filename` arg).

Reply:
After reviewing this was just me being silly, I forgot that InvalidTemplateError was actually a subclass of the other one and was just assuming I had to rescue it as well. Fixing.